### PR TITLE
For 9.0 we don't have the possibility to store calendar and addressbo…

### PR DESCRIPTION
**Pull request by DeepDiver1975**
…ok properties on a per-user basis and therefore we simple don't allow this for now

@georgehrke

### CODE REVIEW

**Comment by georgehrke** [comment]
Will this only affect the properties inside `oc_calendars` and `oc_addressbooks` or also `oc_properties`?

For `oc_properties` the principal for shared calendars is different. (e.g. `calendars/user2/owncloud_shared_by_user1`).

Prohibiting changes made to oc_properties is an absolute no-go, because then would end up with not being able to hide shared calendars yet again.

**Comment by DeepDiver1975** [comment]
> Will this only affect the properties inside oc_calendars and oc_addressbooks

yes

**Comment by georgehrke** [comment]
You said that using `oc_properties` for everything would be too slow.
How about only using it for shared calendars (given that the user changed properties like color, etc.)?

If I'm not completely mistaken this could be pretty easily by achieved by using an empty array for supportedProperties if the calendar is shared.
https://github.com/owncloud/core/blob/master/apps/dav/lib/caldav/caldavbackend.php#L392

**Comment by DeepDiver1975** [comment]
> You said that using oc_properties for everything would be too slow.
> How about only using it for shared calendars (given that the user changed properties like color, etc.)?

I want to retrieve these per-user properies in one sql - bet by simply joining them together.
With the structure of oc_properties this is not possible.

**Comment by georgehrke** [comment]
But doesn't Sabre/DAV query oc_properties anyway? (e.g. for `{http://owncloud.org/ns}calendar-enabled`)

For a shared calendar it would just overwrite the color stored in oc_calendars with the '{http://apple.com/ns/ical/}calendar-color' property from oc_calendars.

**Comment by DeepDiver1975** [comment]
> But doesn't Sabre/DAV query oc_properties anyway? (e.g. for {http://owncloud.org/ns}calendar-enabled)

but only for those props which are requested - so if you request a custom prop - we will get it from oc_properties. But the set of default props we should try to have accessible with one call.

So I'd even suggest to move this enabled prop to the oc_calendar table :wink:

**Comment by georgehrke** [comment]
> So I'd even suggest to move this enabled prop to the oc_calendar table :wink:

Okay, but please please not for 9.0 if sharees can't edit these properties.
Not being able to change calendar name or color is one thing, but not being able to hide it is not acceptable. (this is one of the worst issues we have in the current calendar)

**Comment by georgehrke** [comment]
> (this is one of the worst issues we have in the current calendar)

and people are not happy about it ;)
https://github.com/owncloud/calendar/issues/21

File: apps/dav/lib/carddav/addressbook.php [hunk 1]
```diff
@@ -161,4 +162,13 @@ function delete() {
 		}
 		parent::delete();
 	}
+
+	function propPatch(PropPatch $propPatch) {
+		function propPatch(PropPatch $propPatch) {
```
Comment (nickvergessen) [inline apps/dav/lib/carddav/addressbook.php]: this looks wrong

File: apps/dav/lib/caldav/calendar.php [hunk 2]
```diff
@@ -99,4 +100,11 @@ function delete() {
 		}
 		parent::delete();
 	}
+
+	function propPatch(PropPatch $propPatch) {
+		if (isset($this->calendarInfo['{http://owncloud.org/ns}owner-principal'])) {
```
Comment (nickvergessen) [inline apps/dav/lib/caldav/calendar.php]: this means it's a shared calendar and the current user is not the owner?

**Comment by nickvergessen** [comment]
Tested and works as described.

However I also see that it is not possible to change the visibility of a calendar anymore, which is somewhat bad. Should we allow the prop patch, if it only changes the `calendar-enabled`:
```php
	function propPatch(PropPatch $propPatch) {
		$mutations = $propPatch->getMutations();
		if (isset($this->calendarInfo['{http://owncloud.org/ns}owner-principal']) && (sizeof($mutations) !== 1 || !isset($mutations['{http://owncloud.org/ns}calendar-enabled']))) {
			throw new Forbidden();
		}
		parent::propPatch($propPatch);
	}
```

That works for me, only the calendar UI does not correctly reflect that (it seems to ignore errors on the proppatch call)

**Comment by nickvergessen** [comment]
Rebased and commited

**Comment by nickvergessen** [comment]
Will add a test for the calendar allowed case

**Comment by DeepDiver1975** [comment]
@georgehrke can we get this in? Or do we want to handle this differently?

**Comment by PVince81** [comment]
Any update ? If this isn't in by tomorrow it will have to be moved to 9.1.

**Comment by DeepDiver1975** [comment]
> Any update ? If this isn't in by tomorrow it will have to be moved to 9.1.

I just pinged @georgehrke on irc - he will review asap - THX @PVince81

**Comment by georgehrke** [comment]
- [ ] If the calendar is read-only, the sharee can't set custom props like calendar-enabled.

**Comment by georgehrke** [comment]
looks good to me otherwise

**Comment by nickvergessen** [comment]
Can be fixed with:
```diff
diff --git a/apps/dav/lib/dav/sharing/backend.php b/apps/dav/lib/dav/sharing/backend.php
index 7556706..a514725 100644
--- a/apps/dav/lib/dav/sharing/backend.php
+++ b/apps/dav/lib/dav/sharing/backend.php
@@ -192,6 +192,12 @@ class Backend {
 					'principal' => $share['{' . \OCA\DAV\DAV\Sharing\Plugin::NS_OWNCLOUD . '}principal'],
 					'protected' => true,
 				];
+			} else if ($this->resourceType === 'calendar') {
+				$acl[] = [
+					'privilege' => '{DAV:}write-properties',
+					'principal' => $share['{' . \OCA\DAV\DAV\Sharing\Plugin::NS_OWNCLOUD . '}principal'],
+					'protected' => true,
+				];
 			}
 		}
 		return $acl;
```

Changing events is still not possible because `{DAV:}write-content` is not given, so that seems to be the solution:
```xml
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAVACL\Exception\NeedPrivileges</s:exception>
  <s:message>User did not have the required privileges ({DAV:}write-content) for path "calendars/test1/test_shared_by_admin/ownCloud-fbjy75a0ng9acajsez833h.ics"</s:message>
  <d:need-privileges>
    <d:resource>
      <d:href>/remote.php/dav/calendars/test1/test_shared_by_admin/ownCloud-fbjy75a0ng9acajsez833h.ics</d:href>
      <d:privilege>
        <d:write-content/>
      </d:privilege>
    </d:resource>
  </d:need-privileges>
</d:error>
```

**Comment by DeepDiver1975** [comment]
@nickvergessen THX - let's give it a try - please commit the change and we can test this

**Comment by nickvergessen** [comment]
rebased the branch and commited

**Comment by georgehrke** [comment]
works :+1:
thx :)

**Comment by MorrisJobke** [comment]
> Any update ? If this isn't in by tomorrow it will have to be moved to 9.1.

Move to 9.1

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
@karlitschek @cmonteroluque we should backport this - thx

**Comment by nickvergessen** [comment]
Backport in https://github.com/owncloud/core/pull/22909

**Comment by karlitschek** [comment]
please backport

**Pull request merged**